dataservice/api/genomic_file/models.py

```diff
@@ -30,9 +49,166 @@ class GenomicFile(db.Model, Base):
     file_url = db.Column(db.Text())
     # See link for why md5sum should use uuid type
     # https://dba.stackexchange.com/questions/115271/what-is-the-optimal-data-type-for-an-md5-field
-    md5sum = db.Column(UUID(), unique=True)
+    md5sum = db.Column(UUID(as_uuid=True), unique=True)
     controlled_access = db.Column(db.Boolean())
     sequencing_experiment_id = db.Column(KfId(),
                                          db.ForeignKey(
-                                         'sequencing_experiment.kf_id'),
-                                         nullable=False)
+                                         'sequencing_experiment.kf_id'))
+
+    # Fields used by indexd, but not tracked in the database
+    urls = []
+    rev = None
+    hashes = {}
+    # The metadata property is already used by sqlalchemy
+    _metadata = {}
+    size = None
+
+    def merge_indexd(self):
+        """
+        Gets additional fields from indexd
+        """
+        if current_app.config['INDEXD_URL'] is None:
+            return
+
+        indexd_url = current_app.config['INDEXD_URL']
+        resp = requests.get(indexd_url + self.uuid)
+
+        # File does not exist
+        if resp.status_code == 404:
+            return
+        for prop, v in resp.json().items():
+            if hasattr(self, prop):
+                setattr(self, prop, v)
+
+        if 'metadata' in resp.json():
+            self._metadata = resp.json()['metadata']
+
+
+@event.listens_for(GenomicFile, 'before_insert')
+def register_indexd(mapper, connection, target):
+    """
+    Registers the genomic file with indexd by sending a POST with simple
+    auth to /index/ with a document formatted as:
+
+      {
+        "file_name": "my_file",
+        "hashes": {
+          "md5": "0b7940593044dff8e74380476b2b27a9"
+        },
+        "size": 123,
+        "urls": [
+          "s3://mybucket/mykey/"
+        ],
+        "metadata": {
+          "acls": "phs000000",
+          "kf_id": "PT_00000000"
+        }
+      }
+
+    The response upon successful registry will contain a `did` which will
+    be used as the target's uuid so that it may be joined with the indexd
+    data.
+    """
+    if current_app.config['INDEXD_URL'] is None:
+        return
+
+    req_body = {
+        "file_name": target.file_name,
+        "size": target.size,
+        "form": "object",
+        "hashes": {
+            "md5": str(target.md5sum).replace('-', '')
```

Comment:
Marshmallow deserialization of UUIDs has the hyphens in there. Likely a better way to do this.